### PR TITLE
rkt: Support alternate stage1's via annotation

### LINKED ISSUE #23944: rkt: Support per-pod stage1-image
It would be useful to be able to set the rkt stage1-image on a per-pod basis (overriding the global `--rkt-stage1-image` flag on the kubelet).

A example use-case would be using the [rkt-fly stage1](https://coreos.com/rkt/docs/latest/running-fly-stage1.html) when running a self-hosted kubelet, while using the global stage1 for all other pods.

One option might be to use annotations on the pod which the rkt runtime could parse for stage1-image. This would be somewhat lightweight in that no api-changes would need to be made. Eventually this could be replaced if/when runtime-specific pod configuration was available.

A concern with using annotations is that the stage1 could essentially be running the pod with escalated privileges. It might mean that at the kubelet side, pods with a stage1 annotation would need to be rejected if they did not also contain the `privileged=true` flag (however, the annotations wouldn't be enforceable from the api-server `--allow-privileged` side).

@yifan-gu @philips

### PR DESCRIPTION
**Pull request by euank**
This provides a basic implementation for setting a stage1 on a per-pod
basis via an annotation.

This provides a basic implementation for setting a stage1 on a per-pod
basis via an annotation. See discussion here for how this approach was arrived at: https://github.com/kubernetes/kubernetes/issues/23944#issuecomment-212653776

It's possible this feature should be gated behind additional knobs, such
as a kubelet flag to filter allowed stage1s, or a check akin to what
priviliged gets in the apiserver.
Currently, it checks `AllowPrivileged`, as a means to let people disable
this feature, though overloading it as stage1 and privileged isn't
ideal.

Fixes #23944

Testing done (note, unfortunately done with some additional ./cluster changes merged in):

```
$ cat examples/stage1-fly/fly-me-to-the-moon.yaml
apiVersion: v1
kind: Pod
metadata:
  labels:
    name: exit
  name: exit-fast
  annotations: {"rkt.alpha.kubernetes.io/stage1-name-override": "coreos.com/rkt/stage1-fly:1.3.0"}
spec:
  restartPolicy: Never
  containers:
    - name: exit
      image: busybox
      command: ["sh", "-c", "ps aux"]
$ kubectl create -f examples/stage1-fly
$ ssh core@minion systemctl status -l --no-pager k8s_2f169b2e-c32a-49e9-a5fb-29ae1f6b4783.service
...
failed
...
May 04 23:33:03 minion rkt[2525]: stage0: error writing /etc/rkt-resolv.conf: open /var/lib/rkt/pods/run/2f169b2e-c32a-49e9-a5fb-29ae1f6b4783/stage1/rootfs/etc/rkt-resolv.conf: no such file or directory
...
# Restart kubelet with allow-privileged=false
$ kubectl create -f examples/stage1-fly
$ kubectl describe exit-fast
...
  1m        19s     5   {kubelet euank-e2e-test-minion-dv3u}    spec.containers{exit}   Warning     Failed      Failed to create rkt container with error: cannot make "exit-fast_default(17050ce9-1252-11e6-a52a-42010af00002)": running a custom stage1 requires a privileged security context
....
```

Note as well that the "success" here is rkt spitting out an [error message](https://github.com/coreos/rkt/issues/2141) which indicates that the right stage1 was being used at least.

cc @yifan-gu @aaronlevy

### CODE REVIEW

File: pkg/kubelet/rkt/rkt.go [hunk 1]
```diff
@@ -876,7 +917,7 @@ func (r *Runtime) generateRunCommand(pod *api.Pod, uuid string) (string, error)
 // On success, it will return a string that represents name of the unit file
 // and the runtime pod.
 func (r *Runtime) preparePod(pod *api.Pod, pullSecrets []api.Secret) (string, *kubecontainer.Pod, error) {
-	// Generate the pod manifest from the pod spec.
+	// Generate the appc pod manifest from the k8 pod spec.
```
Comment (yifan-gu) [inline pkg/kubelet/rkt/rkt.go]: s/k8/k8s

Comment (euank) [inline pkg/kubelet/rkt/rkt.go]: ah, I thought the 8 was pronounced `ubernetes`, not `ubernete`. Thanks! :ferris_wheel:

**Comment by yifan-gu** [comment]
LGTM, cc @bgrant0607 @philips @kubernetes/sig-node @pb0

**Comment by yifan-gu** [comment]
Applying LGTM, feel free to remove if you guys have any concerns or questions :)
cc @bgrant0607 @philips @kubernetes/sig-node @pb0

**Comment by bprashanth** [comment]
your failure is https://github.com/kubernetes/kubernetes/issues/25375, but you need a rebase anyway

**Comment by yifan-gu** [comment]
`pkg/kubelet/rkt/rkt.go:986: r.runCommand undefined (type *Runtime has no field or method runCommand, but does have RunCommand)`

**Comment by yifan-gu** [comment]
@k8s-bot ok to test.

**Comment by euank** [comment]
This'll rebase conflict again once #25902 gets in.

**Pull request merged**